### DISCUSSION TITLE
A few more checks for Tools > Check > For Corrupted Files

### DISCUSSION THREAD

**Discussion by CamDawg** [NearInfinityBrowser/NearInfinity#170] [Ideas & Feature Requests]
I typically run most, if not all, of NI's checks when I start maintenance of a new mod, as it saves all sorts of troubles down the line. This came up in the [context of Amber](https://www.gibberlings3.net/forums/topic/37360-version-7-of-amber-now-available/#comment-330290): long story short, a pair of creature files listed offsets as zero. The files worked as the associated counts were also zero, but this could (and in this case did) cause problems down the line as other mods seek to read or add info to these fields. 

To that end, I'd like to propose an additional series of file checks a part of this process. I'd like NI to flag (either as error or warning) any file which has offset fields pointing to the 'static' part of a file type, e.g. in this particular case these would have been flagged as being in the header (value < 0x2d4) of a CRE v1.0 file. 

If it helps here are the lengths of the headers of the current file types available for this check:

- are
  - v1.0: 0x11c
  - v9.1: 0x12c
- chr
  - v1, v1.2, v2, v9: 0x64
  - v2.2: 0x224
- chu: 0x14
- cre
  - v1: 0x2d4
  - v1.2: 0x378
  - v2.2: 0x62e
  - v9: 0x33c 
- dlg: 0x2c (the four-byte field at 0x30 can be omitted AFAIK)
- eff: n/a
- gam
  - v1.1: 0xb8
  - v2.0, v2.2: 0xb4
- itm
  - v1: 0x72
  - v1.1: 0x9a
  - v9: 0x82
- pro: 0x100
- spl
  - v1: 0x72
  - v2: 0x82
- sto
  - v1, v1.1: 0x9c
  - v9: 0xf0
- vef: 0x18
- vvc: n/a
- wed: 0x20
- wmp: 0x10

**Comment by Argent77** [comment]
I can certainly implement such a check. But wouldn't that produce a lot of noise? I can imagine there are many zeroed out offsets in the game resources?

**Comment by Sampsca** [comment]
What's the alternative of zeroed offsets to data blocks that don't exist?  Pointing to the last byte of the file, even if that's pointing into some irrelevant structure?  Pointing beyond the end of the file?  Neither seem any better to me...

**Comment by Argent77** [comment]
I did a first test, and the "noise" is mostly present in ARE resources where almost every unused offset is currently zero.

**Comment by CamDawg** [comment]
> What's the alternative of zeroed offsets to data blocks that don't exist? Pointing to the last byte of the file, even if that's pointing into some irrelevant structure? Pointing beyond the end of the file? Neither seem any better to me...

Usually you point it wherever such effects would normally be inserted. For effects that's typically between the memorization info and the item block in a creature file e.g. The danger here is an edge case, as evidenced by the Amber link above. Basically you need one of these junk offsets plus another mod trying to utilize them--in this case the second mod was intelligent and recognized it as a bad offset, because the alternative would have been inserting effects at the beginning of the file and corrupting it. 

> I did a first test, and the "noise" is mostly present in ARE resources where almost every unused offset is currently zero.

More stuff for the Fixpack, woo!

In seriousness, thanks for checking on this. If it's an overwhelming number of false positives then it severely reduces the utility of the check. It's not currently structured to have the warning/error tabs you get for the dialogue checks, but perhaps that could be an alternative?

**Comment by Argent77** [comment]
> It's not currently structured to have the warning/error tabs you get for the dialogue checks, but perhaps that could be an alternative?

I don't think Warning tabs make sense for checking corrupted files. It's either valid data or corrupted data.

Feature has been implemented, and it confirms my initial tests that only ARE resources produce a greater number of issues. There are also some issues with CHR files. But I don't think there is anything we can do about it since they are created by the game engine itself.

The feature can already be tested in the [Nightly Release](https://github.com/Argent77/NearInfinity/releases/tag/nightly).

**Comment by CamDawg** [comment]
See, I'm over wasting time in Discord while you're quietly knocking this out. Cheers!
